### PR TITLE
Refactor SchemaForm into SchemaForm/TabForm

### PR DESCRIPTION
**Pull request by kennyt**
you can now create a form with tabs on the left and a selected form on the right without having to do anything with a schema.

px

### CODE REVIEW

**Comment by mesosphere-web** [comment]
Reviewer: @aldipower

File: src/js/components/TabForm.js [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+    this.props.getTriggerSubmit(this.handleExternalSubmit);
+  }
+
+  componentDidMount() {
+    this.setState({renderGemini: true});
+  }
+
+  componentDidUpdate() {
+    // Timeout necessary due to modal content height updates on did mount
+    setTimeout(() => {
```
Comment (aldipower) [inline src/js/components/TabForm.js]: You could consider the MutationObserver to listen on render changes with a save way.
https://developer.mozilla.org/de/docs/Web/API/MutationObserver

Just a suggestion, feel free to ignore or implement in another PR.

File: src/js/components/TabForm.js [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+    super();
+
+    this.state = {currentTab: '', renderGemini: false};
+
+    METHODS_TO_BIND.forEach((method) => {
+      this[method] = this[method].bind(this);
+    });
+
+    this.triggerSubmit = function () {};
+    this.isValidated = true;
```
Comment (aldipower) [inline src/js/components/TabForm.js]: Where is this used?

Comment (kennyt) [inline src/js/components/TabForm.js]: nowhere. removed, thanks :)

File: src/js/components/TabForm.js (hunk 1 above)
Comment (kennyt) [inline src/js/components/TabForm.js]: looks cool. i didn't add this originally, maybe we can do it later

**Comment by ssk2** [comment]
retest this please

File: src/js/components/TabForm.js [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+  buildModel() {
+    Object.keys(this.props.definition).forEach((formKey) => {
+      this.submitMap[formKey]();
+    });
+  }
+
+  getTriggerSubmit(formKey, triggerSubmit) {
+    this.submitMap[formKey] = triggerSubmit;
+  }
```
Comment (pierlo-upitup) [inline src/js/components/TabForm.js]: I know this comes from somewhere else, but I really don't like a `get*` method that behaves like a `set` one.

File: src/js/components/TabForm.js [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+
+TabForm.defaultProps = {
+  className: 'multiple-form row',
+  getTriggerSubmit: function () {},
+  onChange: function () {},
+  onSubmit: function () {}
+};
+
+TabForm.propTypes = {
+  className: React.PropTypes.string,
```
Comment (pierlo-upitup) [inline src/js/components/TabForm.js]: Let's have all props listed here. Especially `definition`, which seems to be required for this to work.

**Comment by pierlo-upitup** [comment]
So what does `SchemaForm` give us, now?

**Comment by kennyt** [comment]
@pierlo-upitup schemaform parses a schema and then creates a form definition, which `TabForm` uses. `TabForm` doesn't know anything about schemas

File: src/js/components/TabForm.js (hunk 3 above)
Comment (kennyt) [inline src/js/components/TabForm.js]: yeah a lot of our components that implement `Form` do this. we should change it all at once

File: src/js/components/TabForm.js (hunk 4 above)
Comment (kennyt) [inline src/js/components/TabForm.js]: done

File: src/js/components/TabForm.js [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+  }
+
+  handleExternalSubmit() {
+    return this.validateForm();
+  }
+
+  validateForm() {
+    this.buildModel();
+    this.props.onSubmit(this.model);
+    return this.model;
```
Comment (jfurrow) [inline src/js/components/TabForm.js]: Supernit, do we want a newline above this `return`? I honestly don't care either way :)

**Comment by jfurrow** [comment]
LGTM!

File: src/js/components/TabForm.js [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+  handleTabClick(currentTab) {
+    this.setState({currentTab});
+  }
+
+  handleFormSubmit(formKey, formModel) {
+    this.model[formKey] = formModel;
+  }
+
+  handleExternalSubmit() {
+    return this.validateForm();
```
Comment (MatApple) [inline src/js/components/TabForm.js]: Do we need to have `handleExternalSubmit` and `validateForm`? Also, `validateForm` isn't really doing any validation right?

File: src/js/components/TabForm.js (hunk 5 above)
Comment (MatApple) [inline src/js/components/TabForm.js]: Lets do it :)

File: src/js/components/TabForm.js [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+  getLabel(description, label) {
+    return (
+      <label>
+        <span className="media-object-spacing-wrapper
+          media-object-spacing-narrow">
+          <div className="media-object">
+            <span className="media-object-item">
+              {label}
+            </span>
+            <Tooltip content={description} wrapperClassName="tooltip-wrapper
```
Comment (MatApple) [inline src/js/components/TabForm.js]: Should we put these props all on newlines like we usually do?

File: src/js/components/TabForm.js [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+        render: this.getHeader.bind(this, title, description)
+      }].concat(definition);
+
+      return (
+        <div key={i} className="form-panel">
+          <Form
+            className={panelClassSet}
+            formGroupClass="form-group flush"
+            definition={formDefinition}
+            triggerSubmit={this.getTriggerSubmit.bind(this, formKey)}
```
Comment (MatApple) [inline src/js/components/TabForm.js]: This is already bound in `METHODS_TO_BIND`

File: src/js/components/TabForm.js (hunk 7 above)
Comment (kennyt) [inline src/js/components/TabForm.js]: alright done

File: src/js/components/TabForm.js (hunk 6 above)
Comment (kennyt) [inline src/js/components/TabForm.js]: alright, i've removed validateForm and put it all in handleExternalSubmit

**Pull request merged**